experimental/InvariantTheory/primary_invariants.jl

```diff
@@ -0,0 +1,188 @@
+  R = polynomial_ring(RG)
+  n = length(degrees)
+
+  deg_dict = Dict{fmpz, Int}()
+  for e in degrees[length(invars) + 1:length(invars) + k]
+    deg_dict[e] = get(deg_dict, e, 0) + 1
+  end
+  for degs in Hecke.subsets(Set(keys(deg_dict)))
+    if isempty(invars)
+      I = ideal(R, R())
+    else
+      I = ideal(R, invars)
+    end
+    for e in degs
+      iter = get!(iters, Int(e)) do
+        vector_space_iterator(coefficient_ring(RG), iterate_basis(RG, Int(e)))
+      end
+      I = I + ideal(R, collect_basis(iter))
+    end
+    numbersInds = sum(Int[ deg_dict[e] for e in degs ])
+    if dim(I) > n - length(invars) - numbersInds
+      return false
+    end
+  end
+  return true
+end
+
+function primary_invariants_via_optimal_hsop(RG::InvRing)
+  iters = Dict{Int, VectorSpaceIterator}()
+  k = 1
+  while true
+    degrees = candidates_primary_degrees(RG, k)
+    for ds in degrees
+      invars = elem_type(polynomial_ring(RG))[]
+      b, _ = primary_invariants_via_optimal_hsop!(RG, ds, invars, iters)
+      b && return invars
+    end
+    k += 1
+  end
+end
+
+# Kemper "An Algorithm to Calculate Optimal Homogeneous Systems of Parameters", 1999, [Kem99]
+# Returns a bool b and an integer k.
+# b == true iff primary invariants of the given degrees exist. In this case
+# invars will contain those invariants.
+# k is only needed for recursive calls of the function.
+function primary_invariants_via_optimal_hsop!(RG::InvRing{FldT, GrpT, PolyElemT}, degrees::Vector{fmpz}, invars::Vector{PolyElemT}, iters::Dict{Int, <: VectorSpaceIterator}) where {FldT, GrpT, PolyElemT}
+  k = 0
+  n = length(degrees) - length(invars)
+  R = polynomial_ring(RG)
+
+  if length(invars) != length(degrees)
+    d = degrees[length(invars) + 1]
+
+    iter = get!(iters, Int(d)) do
+      vector_space_iterator(coefficient_ring(RG), iterate_basis(RG, Int(d)))
+    end
+
+    for f in iter
+      if f in invars
+        continue
+      end
+      push!(invars, f)
+      if k > 0
+        if !check_primary_degrees(RG, degrees, invars, k - 1, iters)
+          pop!(invars)
+          continue
+        end
+      end
+      b, kk = primary_invariants_via_optimal_hsop!(RG, degrees, invars, iters)
+      if b
+        return true, 0
+      end
+      pop!(invars)
+      if kk >= k
+        k = kk + 1
+        if !check_primary_degrees(RG, degrees, invars, k, iters)
+          break
+        end
+      end
+    end
+  end
+
+  if k <= 1
+    if !isempty(invars) && dim(ideal(polynomial_ring(RG), invars)) > n
+      k = 0
+    else
+      k = 1
+    end
+  end
+  if iszero(n) && isone(k)
+    return true, k
+  end
+  return false, k
+end
+
+# Shouldn't be here and shouldn't be like this
+order(G::MatrixGroup{T}) where {T <: Union{nf_elem, fmpq}} = order(isomorphic_group_over_finite_field(G)[1])
```

Comment:
If your bottleneck is that you ask for the order of the *same* group over and over again, then you could workaround the overhead issue *for now* by doing something like this (untested):
```suggestion
function order(::Type{T}, G::MatrixGroup{U}) where {T <: IntegerUnion, U <: Union{nf_elem, fmpq}}
   res = get_special(G, :order)
   if res == nothing
     res = order(fmpz, isomorphic_group_over_finite_field(G)[1])
     set_special(G, :order => res)
   end
   return T(res)::T
end
```

Reply:
sorry, previous version missed the type `U` for the matrix group... Anyway, as I said, code is untested -- it should work in principle but may contain syntax error or other silly mistakes on my part.

Reply:
Yeah... I don't want to add workaround over workaround because of this...
Also, it's not just the order for which I need the group isomorphism.